### ISSUE TITLE
[Enhancement]: Add filter for `wc_product_has_unique_sku ` check

### ISSUE THREAD

**Issue by ovidiul** [woothemes/woocommerce#31720]
### Describe the solution you'd like

On a large store with over 500K products, this code [line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-product-functions.php#L618)
```
$sku_found  = $data_store->is_existing_sku( $product_id, $sku );
```
has the potential to generate a slow query as below. 

Inserting a new product always takes at least 5 seconds for that call to finish.

<img width="1507" alt="Screenshot 2022-01-25 at 11 28 40" src="https://user-images.githubusercontent.com/537751/150949657-c774286b-d1f9-45d3-a919-1c0e26110084.png">

### Describe alternatives you've considered

It would be beneficial if an additional conditional filter would be used there so that the SKU check could be done in alternative ways. 

We are looking to leverage ElasticSearch for the slow queries execution, so that filter would allow us to send that query to the ES server.

Something in the lines of

```
if ( apply_filters( 'wc_product_pre_has_unique_sku', false, $product_id, $sku ) ) {
	return true;
}
```

Happy to add a PR if this change could be considered.

**Comment by moory-se** [comment]
Hi,

I recently worked on another slow query from a lookup table (https://github.com/woocommerce/woocommerce/issues/31688). I do think that this is a very similiar case.

The lookup tables are for sure made for performance reasons - after all, they are in fact "lookup tables" and not "storage tables". Given that fact, they should utilize indexes quite heavily. This table (`wp_wc_product_meta_lookup`) does in fact have quite a few indexes, but not on the SKU column.

I can not test this query with 500k products, but I would bet lots of money that the query would be a lot faster with an index on the SKU column. This is the results from this table with 500k "faked" rows. It is only the select part from the `wp_wc_product_meta_lookup` table (i.e not the join with wp_posts, but that is not the problem here).

**Without index**
The query runs at ~120 ms and needs to scan the entire table.
![image](https://user-images.githubusercontent.com/95882730/150985104-f0223468-7c03-4808-88d3-6db3c8430ac7.png)

**With index**
The query only uses the index, and runs at ~0.8 ms. To clarify: this is 150x faster. I'm not entirely sure how indexes works on varchars, but i would guess that the composition of your SKUs would change this number - i.e it could be both faster and slower than this. It would however always be faster than without the index.
![image](https://user-images.githubusercontent.com/95882730/150985308-b7945e51-bc7f-4e7c-90df-70b06be1ab9a.png)

This is clearly a huge improvement already. I however do think the query would be faster if we rewrote the query to utilize a LIMIT 0,1 or EXISTS().
 If you could provide me a database dump of the 2 tables (`wp_wc_product_meta_lookup` and `wp_posts` i could look into it more in detail.

If not, adding the index on the `wp_wc_product_meta_lookup.sku` is a good improvement and should be implemented either way. 

Please note that all InnoDB indexes are prefixed with the primary key (product_id in this case), which is why we don't have to add that column to the index.

**Comment by alexminza** [comment]
Included in milestone 9.0.0 https://github.com/woocommerce/woocommerce/milestone/231?closed=1

* https://github.com/woocommerce/woocommerce/issues/46759
* https://github.com/woocommerce/woocommerce/pull/46763
